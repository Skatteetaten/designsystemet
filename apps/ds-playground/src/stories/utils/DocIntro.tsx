import { ReactNode } from 'react';

import { Link } from '@skatteetaten/ds-buttons';
import { Paragraph } from '@skatteetaten/ds-typography';

interface DocIntroProps {
  name: string;
  url: string;
}

export const DocIntro = ({ name, url }: DocIntroProps): ReactNode => {
  return (
    <Paragraph variant={'ingress'}>
      {`Se `}
      <Link href={url}>{name}</Link>
      {` p√• stil og tone for eksempler og
        retningslinjer for denne komponenten.`}
    </Paragraph>
  );
};
