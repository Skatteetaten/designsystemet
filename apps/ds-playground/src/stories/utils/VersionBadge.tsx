import { ReactNode } from 'react';

interface VersionBadgeProps {
  packageName: string;
}

export const VersionBadge = ({ packageName }: VersionBadgeProps): ReactNode => {
  let color = 'green';

  let version =
    import.meta.env.DS_VERSIONS[`${packageName.replaceAll('-', '_')}`] ?? '..';
  if (import.meta.env.MODE === 'development') {
    color = 'red';
    version = 'developing';
  }
  if (import.meta.env['STORYBOOK_BUILD_VERSION']?.includes('SNAPSHOT')) {
    color = 'blue';
  }

  const normalizedPackageName = packageName.replaceAll('-', '--');
  const normalizedVersion = version.replaceAll('-', '--').replaceAll('_', '__');

  return (
    <img
      src={`https://img.shields.io/badge/%40skatteetaten%2Fds--${normalizedPackageName}-${normalizedVersion}-${color}?logo=npm`}
      alt={`@skatteetaten/ds-${packageName}@${version}`}
    />
  );
};
