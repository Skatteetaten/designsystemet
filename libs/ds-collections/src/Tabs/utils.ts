export const valueRegex = /^[A-Za-z][\w\-.:]*$/;
