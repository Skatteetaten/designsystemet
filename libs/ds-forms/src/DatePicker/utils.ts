import { format, isValid, parse } from 'date-fns';

export const formatDateForInput = (
  dateFormat: string,
  date?: Date | null
): string => {
  return !isValid(date) ? '' : format(date as Date, dateFormat);
};

const COMMON_INPUT_FORMATS = [
  'dd.MM.yy',
  'dd/MM/yy',
  'dd-MM-yy',
  'dd.MM.yyyy',
  'dd/MM/yyyy',
  'dd-MM-yyyy',
  'ddMM',
  'ddMMyy',
  'ddMMyyyy',
];

export const parseDateFromInput = (
  value: string,
  dateFormat: string
): Date | null => {
  const dateFormats = Array.from(
    new Set([...COMMON_INPUT_FORMATS, dateFormat])
  );

  for (const format of dateFormats) {
    const date = parse(value, format, new Date());
    if (isValid(date)) {
      return date;
    }
  }

  return null;
};
