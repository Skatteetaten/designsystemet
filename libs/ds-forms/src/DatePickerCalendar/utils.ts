import { dsI18n } from '@skatteetaten/ds-core-utils';
import {
  addDays,
  getWeekOfMonth,
  isAfter,
  isBefore,
  isSunday,
  isToday,
} from 'date-fns';

const lastValidYear = 9999;
const sunday = 0;

function getDaysFromPrevMonth(datesInSelectedMonth: Date[]): number {
  const firstDayInSelectedMonth = datesInSelectedMonth[0].getDay();
  let daysFromPrevMonthInFirstWeekOfSelectedMonth;
  if (firstDayInSelectedMonth === sunday) {
    daysFromPrevMonthInFirstWeekOfSelectedMonth = 6;
  } else {
    daysFromPrevMonthInFirstWeekOfSelectedMonth = firstDayInSelectedMonth - 1;
  }

  return daysFromPrevMonthInFirstWeekOfSelectedMonth;
}

function getDaysFromNextMonth(datesInSelectedMonth: Date[]): number {
  const lastDayInSelectedMonth =
    datesInSelectedMonth[datesInSelectedMonth.length - 1].getDay();
  let daysFromNextMonthInLastWeekOfSelectedMonth;
  if (lastDayInSelectedMonth === sunday) {
    daysFromNextMonthInLastWeekOfSelectedMonth = 0;
  } else {
    daysFromNextMonthInLastWeekOfSelectedMonth = 7 - lastDayInSelectedMonth;
  }

  return daysFromNextMonthInLastWeekOfSelectedMonth;
}

function getLastDateInPrevMonth(year: number, monthIndex: number): Date {
  const lastDateInPrevMonth = new Date(year, monthIndex, 0);
  lastDateInPrevMonth.setFullYear(lastDateInPrevMonth.getFullYear());

  return lastDateInPrevMonth;
}

function getFirstDateInNextMonth(year: number, monthIndex: number): Date {
  const firstDateInNextMonth = new Date(year, monthIndex + 1, 1);
  firstDateInNextMonth.setFullYear(firstDateInNextMonth.getFullYear());

  return firstDateInNextMonth;
}

interface Cell {
  date: Date;
  isAdjacentMonth: boolean;
  isToday: boolean;
  text: string;
  disabled: boolean;
}

function getCalendarCells(
  year: number,
  monthIndex: number,
  minDate?: Date,
  maxDate?: Date,
  disabledDates?: Date[]
): Cell[] {
  const dates = getDatesInMonth(year, monthIndex);
  const cells: Cell[] = [];

  const daysFromPrevMonth = getDaysFromPrevMonth(dates);
  const lastDateInPrevMonth = getLastDateInPrevMonth(year, monthIndex);

  for (let i = 0; i < daysFromPrevMonth; i++) {
    const date = new Date(lastDateInPrevMonth);

    const cell: Cell = {
      date,
      isAdjacentMonth: true,
      disabled:
        disabledDates?.some((d) => d.getTime() === date.getTime()) ||
        isNotInAllowedRange(date, minDate, maxDate),
      isToday: isToday(date),
      text: String(date.getDate()),
    };

    cells.unshift(cell);

    lastDateInPrevMonth.setDate(lastDateInPrevMonth.getDate() - 1);
  }

  dates.forEach((date) => {
    const cell: Cell = {
      date,
      isAdjacentMonth: false,
      disabled:
        disabledDates?.some((d) => d.getTime() === date.getTime()) ||
        isNotInAllowedRange(date, minDate, maxDate),
      isToday: isToday(date),
      text: String(date.getDate()),
    };

    cells.push(cell);
  });

  const daysFromNextMonth = getDaysFromNextMonth(dates);
  const firstDateInNextMonth = getFirstDateInNextMonth(year, monthIndex);
  for (let i = 0; i < daysFromNextMonth; i++) {
    const date = new Date(firstDateInNextMonth);

    const cell: Cell = {
      date,
      isAdjacentMonth: true,
      disabled:
        disabledDates?.some((d) => d.getTime() === date.getTime()) ||
        isNotInAllowedRange(date, minDate, maxDate),
      isToday: isToday(date),
      text: String(date.getDate()),
    };

    cells.push(cell);

    firstDateInNextMonth.setDate(firstDateInNextMonth.getDate() + 1);
  }

  return cells;
}

function getDatesInMonth(year: number, monthIndex: number): Date[] {
  const date = new Date(year, monthIndex);
  date.setFullYear(year);
  date.setDate(1);

  const month = date.getMonth();
  const dates: Date[] = [];
  while (date.getMonth() === month) {
    dates.push(new Date(date));
    date.setDate(date.getDate() + 1);
    date.setHours(0, 0, 0);
  }
  return dates;
}

export function getCalendarRows(
  year: number | string,
  monthIndex: number,
  minDate?: Date,
  maxDate?: Date,
  disabledDates?: Date[]
): Array<Cell>[] {
  year = findValidYear(year);
  const cells = getCalendarCells(
    year,
    monthIndex,
    minDate,
    maxDate,
    disabledDates
  );
  const rows: Array<Cell>[] = [];
  for (let i = 0; i < cells.length; i += 7) {
    rows.push(cells.slice(i, i + 7));
  }

  return rows;
}

export const isNotInAllowedRange = (
  date: Date,
  minDate?: Date,
  maxDate?: Date
): boolean => {
  minDate && minDate.setHours(0, 0, 0);
  return (
    (minDate ? isBefore(date, minDate) : false) ||
    (maxDate ? isAfter(date, maxDate) : false) ||
    date.getFullYear() > lastValidYear
  );
};

export const getNameOfMonthsAndDays = (): {
  monthNames: string[];
  dayNames: string[];
} => {
  return {
    monthNames: [
      dsI18n.t('ds_forms:datepicker.January'),
      dsI18n.t('ds_forms:datepicker.February'),
      dsI18n.t('ds_forms:datepicker.March'),
      dsI18n.t('ds_forms:datepicker.April'),
      dsI18n.t('ds_forms:datepicker.May'),
      dsI18n.t('ds_forms:datepicker.June'),
      dsI18n.t('ds_forms:datepicker.July'),
      dsI18n.t('ds_forms:datepicker.August'),
      dsI18n.t('ds_forms:datepicker.September'),
      dsI18n.t('ds_forms:datepicker.October'),
      dsI18n.t('ds_forms:datepicker.November'),
      dsI18n.t('ds_forms:datepicker.December'),
    ],
    dayNames: [
      dsI18n.t('ds_forms:datepicker.Mon'),
      dsI18n.t('ds_forms:datepicker.Tue'),
      dsI18n.t('ds_forms:datepicker.Wed'),
      dsI18n.t('ds_forms:datepicker.Thu'),
      dsI18n.t('ds_forms:datepicker.Fri'),
      dsI18n.t('ds_forms:datepicker.Sat'),
      dsI18n.t('ds_forms:datepicker.Sun'),
    ],
  };
};

export function findValidYear(year: string | number): number {
  return year === '' || year === 0 ? new Date().getFullYear() : Number(year);
}

export function getGridIdxForDate(focusableDate: Date): string {
  const colIdx = isSunday(focusableDate) ? 6 : focusableDate.getDay() - 1;
  const rowIdx = getWeekOfMonth(focusableDate, { weekStartsOn: 1 }) - 1;

  return `${rowIdx}${colIdx}`;
}

export const getFirstFocusableDate = (
  selectedDate: Date,
  minDate?: Date,
  maxDate?: Date
): Date => {
  if (isNotInAllowedRange(selectedDate, minDate, maxDate)) {
    let focusableDate = undefined;
    if (maxDate && isAfter(selectedDate, maxDate)) {
      focusableDate = maxDate;
    } else if (minDate && isBefore(selectedDate, minDate)) {
      focusableDate = minDate;
    }

    if (
      focusableDate &&
      !isNotInAllowedRange(focusableDate, minDate, maxDate)
    ) {
      return focusableDate;
    }
  }

  return selectedDate;
};

export const findNextAvailableDate = (
  startDate: Date,
  disabledDates?: Date[]
): Date => {
  if (!disabledDates || disabledDates.length === 0) {
    return addDays(startDate, 1);
  }

  const maxDate = new Date().setFullYear(lastValidYear);
  const disabledTimestamps = new Set(
    disabledDates.map((date) => date.getTime())
  );

  let currentDate = addDays(startDate, 1);
  let currentTimestamp = currentDate.getTime();

  while (
    disabledTimestamps.has(currentTimestamp) &&
    isBefore(currentDate, maxDate)
  ) {
    currentDate = addDays(currentDate, 1);
    currentTimestamp = currentDate.getTime();
  }

  return currentDate;
};

export const findPreviousAvailableDate = (
  startDate: Date,
  disabledDates?: Date[]
): Date => {
  if (!disabledDates || disabledDates.length === 0) {
    return addDays(startDate, -1);
  }

  const minDate = new Date('1000-01-01');
  const disabledTimestamps = new Set(
    disabledDates.map((date) => date.getTime())
  );

  let currentDate = addDays(startDate, -1);
  let currentTimestamp = currentDate.getTime();

  while (
    disabledTimestamps.has(currentTimestamp) &&
    isBefore(minDate, currentDate)
  ) {
    currentDate = addDays(currentDate, -1);
    currentTimestamp = currentDate.getTime();
  }

  return currentDate;
};
