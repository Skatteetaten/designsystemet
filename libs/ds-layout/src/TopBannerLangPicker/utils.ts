import { Languages } from '@skatteetaten/ds-core-utils';

import { TopBannerLangPickerProps } from './TopBannerLangPicker.types';

export const convertLocaleToLang = (locale: Languages): string => {
  return locale.substring(0, locale.indexOf('_'));
};

export type LanguageItem = {
  lang: string;
  displayName: string;
};

export type LanguageItems = {
  [key: string]: LanguageItem;
};

export const isLanguages = (value: unknown): value is Languages => {
  return Object.values(Languages).includes(value as Languages);
};

export const getCurrentLanguages = (
  includeSami: boolean,
  additionalLanguages: TopBannerLangPickerProps['additionalLanguages'] = []
): LanguageItems => {
  const defaultLanguages: LanguageItems = {
    nb: {
      lang: 'nb',
      displayName: 'Bokmål',
    },
    nn: {
      lang: 'nn',
      displayName: 'Nynorsk',
    },
    en: {
      lang: 'en',
      displayName: 'English',
    },
  };
  const se = {
    se: {
      lang: 'se',
      displayName: 'Sámegiella',
    },
  };

  const additionalLanguagesAsObject =
    additionalLanguages?.reduce<LanguageItems>((acc, curr) => {
      acc[curr.lang] = curr;
      return acc;
    }, {});

  return includeSami
    ? { ...defaultLanguages, ...se, ...additionalLanguagesAsObject }
    : { ...defaultLanguages, ...additionalLanguagesAsObject };
};
