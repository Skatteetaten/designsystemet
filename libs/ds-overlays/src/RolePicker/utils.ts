import { formatter } from '@skatteetaten/ds-core-utils';

export const formatOrganisationNumber = (
  organisationNumber: string
): string => {
  const result = formatter({
    type: 'organisationNumber',
    value: organisationNumber,
  });

  return result.value;
};

export const formatNationalIdentityNumber = (
  nationalIdentityNumber: string
): string => {
  const result = formatter({
    type: 'nationalIdentityNumber',
    value: nationalIdentityNumber,
  });

  return result.value;
};
