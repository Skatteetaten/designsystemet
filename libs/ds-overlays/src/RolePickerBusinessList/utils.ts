import { dsI18n } from '@skatteetaten/ds-core-utils';

import { Business } from '../RolePicker/RolePicker.types';

const unitTypesToHide = [
  'AS',
  'DA',
  'ANS',
  'ASA',
  'SE',
  'BA',
  'SA',
  'SCE',
  'STI',
  'IKS',
  'KF',
  'FKF',
  'RHF',
  'HF',
  'NUF',
];

export const getBusinessTitle = (business: Business): string => {
  let title = business.name;
  if (!unitTypesToHide.includes(business.unitType)) {
    title += ` ${business.unitType}`;
  }

  if (business.isDeleted) {
    title += ` (${dsI18n.t('ds_overlays:rolepicker.Deleted')})`;
  }

  return title;
};
