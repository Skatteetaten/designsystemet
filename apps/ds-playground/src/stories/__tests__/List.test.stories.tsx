import { StoryFn, Meta, StoryObj } from '@storybook/react';
import { expect, within } from '@storybook/test';

import {
  List,
  listAsArr,
  ListProps,
  Paragraph,
} from '@skatteetaten/ds-typography';

import { wrapper } from './testUtils/storybook.testing.utils';

const meta = {
  component: List,
  title: 'Tester/List/List',
  argTypes: {
    // Baseprops
    ref: { table: { disable: true } },
    className: { table: { disable: true } },
    id: { table: { disable: true } },
    lang: { table: { disable: true } },
    'data-testid': { table: { disable: true } },
    // Props
    canBeManuallyFocused: { table: { disable: true } },
    children: {
      table: { disable: true },
    },
    hasSpacing: { table: { disable: true } },
    // Tags
    as: {
      table: { disable: true },
      options: [...listAsArr],
      control: 'inline-radio',
    },
  },
} satisfies Meta<typeof List>;
export default meta;
type Story = StoryObj<typeof meta>;

const defaultArgs: ListProps = {
  children: [
    <List.Element key={'listElement_1'}>
      {'Kjenner du behovet til brukeren?'}
    </List.Element>,
    <List.Element key={'listElement_2'}>
      {'Er du sikker på at du kjenner behovet til brukeren?'}
    </List.Element>,
    <List.Element key={'listElement_3'}>{'Snakk med andre.'}</List.Element>,
  ],
};

export const WithRef = {
  name: 'With Ref (FA1)',
  args: {
    ...defaultArgs,
    ref: (instance: HTMLUListElement | null): void => {
      if (instance) {
        instance.id = 'dummyIdForwardedFromRef';
      }
    },
  },
  argTypes: {
    ref: { table: { disable: false } },
  },
  parameters: {
    imageSnapshot: { disable: true },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const list = canvas.getByRole('list');
    await expect(list).toHaveAttribute('id', 'dummyIdForwardedFromRef');
  },
} satisfies Story;

export const WithAttributes = {
  name: 'With Attributes (FA2-5)',
  args: {
    ...defaultArgs,
    id: 'htmlid',
    className: 'dummyClassname',
    lang: 'nb',
    'data-testid': '123ID',
  },
  argTypes: {
    id: { table: { disable: false } },
    className: { table: { disable: false } },
    lang: { table: { disable: false } },
    'data-testid': { table: { disable: false } },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const list = canvas.getByRole('list');
    await expect(list).toHaveAttribute('id', 'htmlid');
    await expect(list).toHaveClass('dummyClassname');
    await expect(list).toHaveAttribute('lang', 'nb');
    await expect(list).toHaveAttribute('data-testid', '123ID');
  },
} satisfies Story;

export const Defaults = {
  name: 'Defaults Variant Bullet (A1, B1, B2)',
  args: {
    ...defaultArgs,
  },
  argTypes: {
    children: {
      table: { disable: false },
      control: { disable: true },
    },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const list = canvas.getByRole('list');
    await expect(list).toBeInTheDocument();
    await expect(list.tagName).toBe('UL');
  },
} satisfies Story;

export const VariantNumber = {
  name: 'Variant Number(A1, B1)',
  args: {
    ...defaultArgs,
    as: 'ol',
  },
  argTypes: {
    as: {
      table: { disable: false },
    },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const list = canvas.getByRole('list');
    await expect(list).toBeInTheDocument();
    await expect(list.tagName).toBe('OL');
  },
} satisfies Story;

export const WithLongTextAndBreaking = {
  name: 'With Long Text And Breaking (A1)',
  args: {
    ...defaultArgs,
    children: [
      <List.Element key={'listElement_1'}>
        {'Denne listItem har en veldig lang tekst. Så lang at den lange teksten tvinger fram linjeskift med ' +
          'tekst som alltid er venstrejustert, brekke over flere linjer og får et hengende innrykk.'}
      </List.Element>,
      <List.Element key={'listElement_2'}>
        {
          'DennelistItemharenveldiglangtekst.Sålangatdentvingerframlinjeskiftmedtekstsomalltidervenstrejustertbrekkeoverflerelinjerogfårethengendeinntrykk.'
        }
      </List.Element>,
    ],
  },
  argTypes: {
    children: {
      table: { disable: false },
      control: { disable: true },
    },
  },
} satisfies Story;

export const WithMarkup = {
  name: 'With Markup (A2, B2)',
  args: {
    ...defaultArgs,
    children: [
      <List.Element key={'listElement_1'}>
        {'Kjenner du behovet til '}
        <strong>{'brukeren?'}</strong>
      </List.Element>,
      <List.Element key={'listElement_2'}>
        {'Behovet til brukeren er veldig viktig. '}
        <a href={'#storybook-root'}>
          {'Er du sikker på at du kjenner behovet til brukeren?'}
        </a>
      </List.Element>,
      <List.Element key={'listElement_3'}>
        {'Snakk med andre om '}
        <em>{'italic, '}</em>
        <mark>{'mark og '}</mark>
        <code>{'code blokk'}</code>
        {' bare på gøy'}
      </List.Element>,
    ],
  },
  argTypes: {
    children: {
      table: { disable: false },
      control: { disable: true },
    },
  },
  parameters: {
    imageSnapshot: {
      hover: `${wrapper} > ul > li a`,
      focus: `${wrapper} > ul > li a`,
    },
  },
} satisfies Story;

const TemplateWithTwoParagraph: StoryFn<typeof List> = (args) => (
  <>
    <Paragraph hasSpacing>
      {'Jeg er et tekstavsnitt uten innrykk. Denne teksten skulle bli ca. to linjer lang, men jeg trenger litt mer innhold for å bli to linjer lang. ' +
        'Derfor fortsetter jeg å skrive på dette tekstavsnittet uten innrykk og etter en stund blir denne teksten ca to linjer lang.'}
    </Paragraph>
    <List {...args} />
    <Paragraph>
      {'Jeg er et tekstavsnitt uten innrykk. Denne teksten skulle bli ca. to linjer lang, men jeg trenger litt mer innhold for å bli to linjer lang. ' +
        'Derfor fortsetter jeg å skrive på dette tekstavsnittet uten innrykk og etter en stund blir denne teksten ca to linjer lang.'}
    </Paragraph>
  </>
);

export const WithSpacing = {
  render: TemplateWithTwoParagraph,
  name: 'With Spacing (A3)',
  args: {
    ...defaultArgs,
    hasSpacing: true,
  },
  argTypes: {
    hasSpacing: { table: { disable: false } },
  },
} satisfies Story;

const TemplateWithVariantsAndAtLeast10ItemsAndOtherComponents: StoryFn<
  typeof List
> = (args) => (
  <>
    <Paragraph>
      {
        'Her kan du se hvordan de forskjellige variantene til en list ser ut sammen med andre elementer med veksling av luft under listene.'
      }
    </Paragraph>
    <List {...args}>
      <List.Element key={'listElement_1'}>
        {'Kjenner du behovet til brukeren?'}
      </List.Element>
      <List.Element key={'listElement_2'}>
        {'Er du sikker på at du kjenner behovet til brukeren?'}
      </List.Element>
      <List.Element key={'listElement_3'}>{'Snakk med andre.'}</List.Element>
      <List.Element key={'listElement_4'}>{'Kjenner du behovet?'}</List.Element>
    </List>
    <Paragraph>
      {'Den første listen kan veksle mellom variantene en liste har.'}
    </Paragraph>
    <List {...args} as={'ol'}>
      <List.Element key={'listElement_1'}>
        {'Kjenner du behovet til brukeren?'}
      </List.Element>
      <List.Element key={'listElement_2'}>
        {'Er du sikker på at du kjenner behovet til brukeren?'}
      </List.Element>
      <List.Element key={'listElement_3'}>{'Snakk med andre'}</List.Element>
      <List.Element key={'listElement_4'}>{'Kjenner du behovet?'}</List.Element>
      <List.Element key={'listElement_5'}>
        {'Er du sikker på at du kjenner behovet til brukeren?'}
      </List.Element>
      <List.Element key={'listElement_6'}>
        {'Snakk med andre når som helst'}
      </List.Element>
      <List.Element key={'listElement_7'}>{'Kjenner du behovet?'}</List.Element>
      <List.Element key={'listElement_8'}>
        {'Er du sikker på at du kjenner behovet til brukeren?'}
      </List.Element>
      <List.Element key={'listElement_9'}>
        {'Snakk med andre ofte'}
      </List.Element>
      <List.Element key={'listElement_10'}>
        {'Snakk med andre aldri'}
      </List.Element>
    </List>
  </>
);

export const WithBothVariantsAndAtLeast10NumberItems = {
  render: TemplateWithVariantsAndAtLeast10ItemsAndOtherComponents,
  name: 'With Both Variants And At Least 10 Number Items',
  args: {
    ...defaultArgs,
  },
} satisfies Story;

export const WithCanBeManuallyFocused: Story = {
  name: 'With Can Receive Focus',
  args: {
    ...defaultArgs,
    canBeManuallyFocused: true,
  },
  argTypes: {
    canBeManuallyFocused: { table: { disable: false } },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const list = canvas.getByRole('list');
    list.focus();
    await expect(list).toBeInTheDocument();
    await expect(list).toHaveAttribute('tabIndex', '-1');
  },
} satisfies Story;
