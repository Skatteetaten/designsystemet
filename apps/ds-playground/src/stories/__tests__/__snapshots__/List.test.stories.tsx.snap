// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tester/List/List Defaults play-test 1`] = `
<div data-test-block="true">
  <ul class="_list_1xbs1_1 _list_withBullet_1xbs1_8  ">
    <li class="_element_1wu1m_1 ">
      Kjenner du behovet til brukeren?
    </li>
    <li class="_element_1wu1m_1 ">
      Er du sikker på at du kjenner behovet til brukeren?
    </li>
    <li class="_element_1wu1m_1 ">
      Snakk med andre.
    </li>
  </ul>
</div>
`;

exports[`Tester/List/List VariantNumber play-test 1`] = `
<div data-test-block="true">
  <ol class="_list_1xbs1_1   ">
    <li class="_element_1wu1m_1 ">
      Kjenner du behovet til brukeren?
    </li>
    <li class="_element_1wu1m_1 ">
      Er du sikker på at du kjenner behovet til brukeren?
    </li>
    <li class="_element_1wu1m_1 ">
      Snakk med andre.
    </li>
  </ol>
</div>
`;

exports[`Tester/List/List WithAttributes play-test 1`] = `
<div data-test-block="true">
  <ul id="htmlid"
      class="_list_1xbs1_1 _list_withBullet_1xbs1_8  dummyClassname"
      lang="nb"
      data-testid="123ID"
  >
    <li class="_element_1wu1m_1 ">
      Kjenner du behovet til brukeren?
    </li>
    <li class="_element_1wu1m_1 ">
      Er du sikker på at du kjenner behovet til brukeren?
    </li>
    <li class="_element_1wu1m_1 ">
      Snakk med andre.
    </li>
  </ul>
</div>
`;

exports[`Tester/List/List WithBothVariantsAndAtLeast10NumberItems smoke-test 1`] = `
<div data-test-block="true">
  <p class="_paragraph_32eqi_1 _paragraph_standard_32eqi_17  ">
    Her kan du se hvordan de forskjellige variantene til en list ser ut sammen med andre elementer med veksling av luft under listene.
  </p>
  <ul class="_list_1xbs1_1 _list_withBullet_1xbs1_8  ">
    <li class="_element_1wu1m_1 ">
      Kjenner du behovet til brukeren?
    </li>
    <li class="_element_1wu1m_1 ">
      Er du sikker på at du kjenner behovet til brukeren?
    </li>
    <li class="_element_1wu1m_1 ">
      Snakk med andre.
    </li>
    <li class="_element_1wu1m_1 ">
      Kjenner du behovet?
    </li>
  </ul>
  <p class="_paragraph_32eqi_1 _paragraph_standard_32eqi_17  ">
    Den første listen kan veksle mellom variantene en liste har.
  </p>
  <ol class="_list_1xbs1_1   ">
    <li class="_element_1wu1m_1 ">
      Kjenner du behovet til brukeren?
    </li>
    <li class="_element_1wu1m_1 ">
      Er du sikker på at du kjenner behovet til brukeren?
    </li>
    <li class="_element_1wu1m_1 ">
      Snakk med andre
    </li>
    <li class="_element_1wu1m_1 ">
      Kjenner du behovet?
    </li>
    <li class="_element_1wu1m_1 ">
      Er du sikker på at du kjenner behovet til brukeren?
    </li>
    <li class="_element_1wu1m_1 ">
      Snakk med andre når som helst
    </li>
    <li class="_element_1wu1m_1 ">
      Kjenner du behovet?
    </li>
    <li class="_element_1wu1m_1 ">
      Er du sikker på at du kjenner behovet til brukeren?
    </li>
    <li class="_element_1wu1m_1 ">
      Snakk med andre ofte
    </li>
    <li class="_element_1wu1m_1 ">
      Snakk med andre aldri
    </li>
  </ol>
</div>
`;

exports[`Tester/List/List WithCanBeManuallyFocused play-test 1`] = `
<div data-test-block="true">
  <ul class="_list_1xbs1_1 _list_withBullet_1xbs1_8  "
      tabindex="-1"
  >
    <li class="_element_1wu1m_1 ">
      Kjenner du behovet til brukeren?
    </li>
    <li class="_element_1wu1m_1 ">
      Er du sikker på at du kjenner behovet til brukeren?
    </li>
    <li class="_element_1wu1m_1 ">
      Snakk med andre.
    </li>
  </ul>
</div>
`;

exports[`Tester/List/List WithLongTextAndBreaking smoke-test 1`] = `
<div data-test-block="true">
  <ul class="_list_1xbs1_1 _list_withBullet_1xbs1_8  ">
    <li class="_element_1wu1m_1 ">
      Denne listItem har en veldig lang tekst. Så lang at den lange teksten tvinger fram linjeskift med tekst som alltid er venstrejustert, brekke over flere linjer og får et hengende innrykk.
    </li>
    <li class="_element_1wu1m_1 ">
      DennelistItemharenveldiglangtekst.Sålangatdentvingerframlinjeskiftmedtekstsomalltidervenstrejustertbrekkeoverflerelinjerogfårethengendeinntrykk.
    </li>
  </ul>
</div>
`;

exports[`Tester/List/List WithMarkup smoke-test 1`] = `
<div data-test-block="true">
  <ul class="_list_1xbs1_1 _list_withBullet_1xbs1_8  ">
    <li class="_element_1wu1m_1 ">
      Kjenner du behovet til
      <strong>
        brukeren?
      </strong>
    </li>
    <li class="_element_1wu1m_1 ">
      Behovet til brukeren er veldig viktig.
      <a href="#storybook-root">
        Er du sikker på at du kjenner behovet til brukeren?
      </a>
    </li>
    <li class="_element_1wu1m_1 ">
      Snakk med andre om
      <em>
        italic,
      </em>
      <mark>
        mark og
      </mark>
      <code>
        code blokk
      </code>
      bare på gøy
    </li>
  </ul>
</div>
`;

exports[`Tester/List/List WithRef play-test 1`] = `
<div data-test-block="true">
  <ul class="_list_1xbs1_1 _list_withBullet_1xbs1_8  "
      id="dummyIdForwardedFromRef"
  >
    <li class="_element_1wu1m_1 ">
      Kjenner du behovet til brukeren?
    </li>
    <li class="_element_1wu1m_1 ">
      Er du sikker på at du kjenner behovet til brukeren?
    </li>
    <li class="_element_1wu1m_1 ">
      Snakk med andre.
    </li>
  </ul>
</div>
`;

exports[`Tester/List/List WithSpacing smoke-test 1`] = `
<div data-test-block="true">
  <p class="_paragraph_32eqi_1 _paragraph_standard_32eqi_17 _paragraph_hasSpacingLarge_32eqi_23 ">
    Jeg er et tekstavsnitt uten innrykk. Denne teksten skulle bli ca. to linjer lang, men jeg trenger litt mer innhold for å bli to linjer lang. Derfor fortsetter jeg å skrive på dette tekstavsnittet uten innrykk og etter en stund blir denne teksten ca to linjer lang.
  </p>
  <ul class="_list_1xbs1_1 _list_withBullet_1xbs1_8 _list_hasSpacing_1xbs1_11 ">
    <li class="_element_1wu1m_1 ">
      Kjenner du behovet til brukeren?
    </li>
    <li class="_element_1wu1m_1 ">
      Er du sikker på at du kjenner behovet til brukeren?
    </li>
    <li class="_element_1wu1m_1 ">
      Snakk med andre.
    </li>
  </ul>
  <p class="_paragraph_32eqi_1 _paragraph_standard_32eqi_17  ">
    Jeg er et tekstavsnitt uten innrykk. Denne teksten skulle bli ca. to linjer lang, men jeg trenger litt mer innhold for å bli to linjer lang. Derfor fortsetter jeg å skrive på dette tekstavsnittet uten innrykk og etter en stund blir denne teksten ca to linjer lang.
  </p>
</div>
`;
